### ISSUE TITLE
Is a sample part of a mutation profile if no mutations were detected?

### ISSUE THREAD

**Issue by sheridancbio** [cBioPortal/cbioportal#3067]
This discussion arises out of issue #2556  : 
- samples are assayed with targeted sequencing for a panel of genes
- some samples have no detected mutation events in any gene of the panel
- those samples are not recorded as part of the mutation genetic_profile for the study
- those samples show up as 'Whole-Exome Sequencing' in the oncoprint

@mandawilson identified the essential issue here : 

**Because these samples are not recorded in the mutation profile, the gene panel API fails to report them when the frontend queries for gene panels by genetic profile stable_id**

I will add additional details below for the reported example case (from the genie_public study), but  to fix this problem it first is important to know whether the definition of genetic_profile in our system is : 
1. all samples which contain a detected event of some type (EXTENDED_MUTATION, CNA, etc..), or

1. all samples which were assayed for events of some type, even if no events were detected

If the answer is (2), then the solution to the oncoprint 'Whole-Exome Sequencing' bug is to adjust the importer so that if we know that a sample was sequenced on *any* gene panel platform (or was sequenced using Whole-Exome or Whole-Genome sequencing if there is a way for us to know that) then it would be recorded as part of the genetic_profile by adding a record into the sample_profile table linking it to the gene panel used. In the current database organization this is the only place to record the gene panel. However, this table is also used as the basis for knowing which samples are part of a genetic profile. So adding records for non-mutated samples which were run using a gene panel would be increasing the size of the profile, which could (I have not traced the code far enough) have subtle effects such as changing the denominator in the calculation of statistics such as "BRCA2 Somatic Mutation Frequency : 3.5%". However, perhaps that change would made the statistic more accurate.

[One side note : sequences which actually were run through whole genome or whole exome sequencing but nonetheless had no detected mutations may or may not be recognizable. Currently we do *not* capture the fact that a sample was run with whole exome sequencing. Perhaps it would be extremely unlikely for such cases to occur anyway.]

If the answer is (1), then we will need another solution; possibly:
- separating the sample_profile table into two tables (one linking samples to profiles, one linking samples to gene_panels)
- adding a special-purpose table of samples which have no mutations but were run through a gene panel
- using the case-list mechanism to capture these samples
- using sample_profile to store all samples which were run on any gene_panel (as above) but recoding the backend to report profile sample lists and list sizes only for samples which had at least one detected mutation

--------------

Details for the example case (GENIE-MSK-P-0001299 in the genie public portal (see any oncoprint - this sample shows as "Whole-Genome Sequencing" even though it was processed with MSK-IMPACT341 )

Database: cgds_genie
cancer_study: genie_public
profiles:
profile: 86 (genie_public_cna)
profile: 87 (genie_public_mutations)
profile: 88 (genie_public_fusion)
<pre>
mysql> select * from sample join sample_profile on sample.internal_id = sample_profile.sample_id where genetic_profile_id = 87 and sample.stable_id like "GENIE-MSK-P-000129%";
+-------------+-----------------------------+---------------------+------------+-------------------+-----------+--------------------+----------+
| INTERNAL_ID | STABLE_ID                   | SAMPLE_TYPE         | PATIENT_ID | TYPE_OF_CANCER_ID | SAMPLE_ID | GENETIC_PROFILE_ID | PANEL_ID |
+-------------+-----------------------------+---------------------+------------+-------------------+-----------+--------------------+----------+
|      483197 | GENIE-MSK-P-0001290-T01-IM3 | Primary Solid Tumor |     472006 | mixed             |    483197 |                 87 |       55 |
|      483198 | GENIE-MSK-P-0001292-T01-IM3 | Primary Solid Tumor |     472007 | mixed             |    483198 |                 87 |       55 |
|      483199 | GENIE-MSK-P-0001293-T01-IM3 | Primary Solid Tumor |     472008 | mixed             |    483199 |                 87 |       55 |
|      483200 | GENIE-MSK-P-0001294-T01-IM3 | Primary Solid Tumor |     472009 | mixed             |    483200 |                 87 |       55 |
|      483201 | GENIE-MSK-P-0001295-T01-IM3 | Primary Solid Tumor |     472010 | mixed             |    483201 |                 87 |       55 |
|      483202 | GENIE-MSK-P-0001296-T01-IM3 | Primary Solid Tumor |     472011 | mixed             |    483202 |                 87 |       55 |
|      483203 | GENIE-MSK-P-0001296-T02-IM5 | Primary Solid Tumor |     472011 | mixed             |    483203 |                 87 |       56 |
|      483204 | GENIE-MSK-P-0001297-T01-IM3 | Primary Solid Tumor |     472012 | mixed             |    483204 |                 87 |       55 |
|      483205 | GENIE-MSK-P-0001298-T01-IM3 | Primary Solid Tumor |     472013 | mixed             |    483205 |                 87 |       55 |
+-------------+-----------------------------+---------------------+------------+-------------------+-----------+--------------------+----------+
9 rows in set (0.03 sec)

mysql> select count(*) from sample_profile where genetic_profile_id = 87;
+----------+
| count(*) |
+----------+
|    17005 |
+----------+
1 row in set (0.01 sec)

mysql> select * from gene_panel where internal_id = 55;
+-------------+---------------+--------------------------------------+
| INTERNAL_ID | STABLE_ID     | DESCRIPTION                          |
+-------------+---------------+--------------------------------------+
|          55 | MSK-IMPACT341 | MSK-IMPACT341, Number of Genes - 341 |
+-------------+---------------+--------------------------------------+
1 row in set (0.00 sec)

</pre>

The count of entries in sample_profile is 17005, but you can see on the portal page and in the staging file showing gene panel per sample that there are 18804 samples.

Please add your comments / questions and let's capture a discussion / decision here.

**Comment by sheridancbio** [comment]
I've been tracing the data paths which read data from sample_profile. Here is my incomplete summary:

**Comment by schultzn** [comment]
I thought we had solved this in the past. If a sample was sequenced (whole exome, targeted, any platform) but has no mutations, the sample is added to the MAF file header of sequenced samples, which includes all samples that were sequenced. This is essential for targeted sequencing panels, as quite a few samples have no mutations. Occasionally, this may affect WXS samples - but I don't think we ever implemented this for any TCGA studies or other WXS projects.

**Comment by pieterlukasse** [comment]
I am pretty sure answer is 1...will look up some sources.

Regarding @schultzn comment: I think we moved this sample list from the MAF header to the mutations meta file some time ago...let me look this up as well.

**Comment by pieterlukasse** [comment]
@sheridancbio so, regarding 1 or 2, in case of mutation data here is some more background https://github.com/cBioPortal/cbioportal/issues/494. Not sure how much has changed since then. 

Issue 494 also discusses a solution for putting the sample list in the meta file, implemented here for the validator https://github.com/cBioPortal/cbioportal/commit/5c803c4e8b4216d97658182de74487a0453a3334, but apparently not used (anymore?) by the loader. Interestingly, the commit does not include changes in the java code...will check more later.

**Comment by pieterlukasse** [comment]
Regarding @schultzn's comment: we moved this sample list from the MAF header to the mutations meta file some time ago (discussed here: https://github.com/cBioPortal/cbioportal/issues/494#issuecomment-183393271 & implemented here for the validation https://github.com/cBioPortal/cbioportal/commit/5c803c4e8b4216d97658182de74487a0453a3334).  Interestingly, the commit does not include changes in the java code. I checked back the history and I was not able to find any implementation in java that reads out the extra header @schultzn mentioned. I only found one unused variable that hints at this functionality (https://github.com/cBioPortal/cbioportal/blob/31d2fdd1d8a6c40c6eca3c81e416b258e09193bf/core/src/main/java/org/mskcc/cbio/cgds/scripts/ImportExtendedMutationData.java#L117...but I can't go back further in history to check). Maybe this was/is in some specific MSK data loading code?

**Comment by pieterlukasse** [comment]
Regarding 1 or 2, the current implementation is 1 for mutations: only samples which contain a detected event of some type are added to the profile (`sample_profile` table). See [#L415](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java#L415) and [#L418](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java#L418). 

I think we should fix this. All profiled samples should be part of `sample_profile`. For mutation data, this means that the extra meta field should be used in the loader code. In the API response and in the frontend we should distinguish between 3 cases for any sample x gene combination: (1) sample was not profiled, (2) sample profiled but contains no detected mutations for the queried gene, (3) sample was profiled and mutations were detected for the queried gene.

**Comment by jjgao** [comment]
It should be #2: if a sample is sequenced, it is sequenced no matter if there is a mutation detected or not.

How are we generating the case list: cases_sequenced?

**Comment by sandertan** [comment]
The importing of gene panel data can be done using 3 methods (described [here](https://github.com/cBioPortal/cbioportal/blob/rc/docs/File-Formats.md#meta-file-13)):
1. gene panel sample matrix
2. gene_panel field in meta file
3. clinical sample file

I think by using method (1) and (3) it is possible to assign samples to a gene panel even if there are no mutations detected. Method (2) only assigns a gene panel to the samples in the MAF file.

Currently there is a bug when only using (1) https://github.com/cBioPortal/cbioportal/issues/3191. But I think using method (1) or (3) would be the solution @sheridancbio describes for the OncoPrint 'Whole-Exome Sequencing' bug.

**Comment by jjgao** [comment]
@sheridancbio do you know what's the status of this?

**Comment by sheridancbio** [comment]
@jjgao asked me to review the status of this discussion and see if there were any action items remaining.
- First, I remember that re-importing the study (and maybe restarting the tomcat web container) resolved the observed problem. (counts showed up correctly)
- The code for processing the header "#sequenced samples: sampleid1 sampleid2 sampleid3 ..." appears to be located in the cbioportal codebase here: org.mskcc.cbio.portal.scripts.ImportExtendedMutationData.processMAFHeader() .. in this function the MAF header is read, and SampleProfile records are prepared for storage through DaoSampleProfile. This defines which samples belong to the (mutation) genetic profile, even if the sample has no mutations in the MAF. I don't remember if the development of this sample_profile table took place before or after this discussion began (I suspect after), but in any case I think that so long as a record places the sample in the profile (perhaps with an identified gene panel id), then when visualizing the profile in the portal the sample should be accounted as a sequenced sample by the web API. I am not familiar with the code for processing the sequenced sample list out of the metadata header; looking through the codebase for the mskcc importer is confusing with a lot of legacy code. There are references to a sequenced_samples.txt file, but I have a feeling that this is a dropped feature. It appears that we have utilities for writing the proper "#sequenced samples:" header to the MAF file in the msk importer, such as ImpactMutationStagingFileRefactoringUtility.initializeStagingFileOutputList(), which takes a Set of sample ids and outputs the header from it, but it seems to be getting the sample ids from the data records themselves. I could do some testing to see what the behavior is. If it is true that a header in the meta_mutations_extended.txt file can control the list of sequenced samples, then send me a spec and I will see if our importer respects that. I don't see it mentioned in https://cbioportal.readthedocs.io/en/latest/File-Formats.html#mutation-data  ... but the docs page also does not mention the "#sequenced samples:" header in the data file either. If this is missing functionality, let's address it in a new issue. For now I think the definition of the samples which were sequenced is provided in the data (MAF) file header.
- With that understanding (MAF file header defines the sequenced sample list), I think we can say that we are using definition #2 above (profile is made up of all samples which were assayed even if no mutations were detected)
- Last question was how the cases_sequenced case_list is created. Of course this could be manually constructed from the information at the top of the MAF files. The MSK importer will either take a manually provided file and parse it, or (more usually) will auto generate the 4 required case lists (all, sequenced, cna, cnaseq) using org.mskcc.cbio.importer.persistence.staging.CaseListFileHandler. cases_sequenced.txt is generated when the function loadCancerStudyStagingFiles() calls importCaseLists() which looks for missing required case lists and generates them with a call to fileUtils.generateCaseLists(). This is somewhat complicated because there can be multiple files that need to be merged (union) or intersected. But I think that a simple mutation (MAF) file will get processed for making the cases_sequenced.txt by function getCaseListFromStagingFile() which first checks whether a mutation file has an associated sequencedSamplesFile (from meta file I presume) and reads the sample set from that file if it exists. Otherwise it reads the MAF file and looks for the MAF header "#sequenced samples:" and returns that list if is is present. Otherwise it looks for a column in the MAF containing the sample Ids and collects all unique values in that column to be returned as the case list. I don't think this logic for generating the case lists from the header in the MAF file is present in the main cBioPortal codebase. I think in general the case lists must be correctly constructed and provided by the person running the import of cancer studies into the portal.

So I think we can close this discussion, but we can open up follow up issues if desired (such as a desire to add features/logic to the cBioPortal core scripts importer which searches the meta file for a supplied filename containing lists of sequenced samples) I'll leave the issue open for now in case I have not correctly captured the msk importer logic above ... maybe @angelicaochoa or @n1zea144 want to comment or make corrections. Let's close this on Nov 15 otherwise.

**Comment by sandertan** [comment]
@jjgao Some things to consider:
1. I recently added validation that if the `_sequenced` case list is missing, an error is reported to make sure data is correct. We can add that having a `#sequenced samples` line is also a valid option.

2. Have you considered adding a property for sequenced samples to the meta file? 

3. The gene panel matrix file has overlapping functionality with the `_sequenced` case list. I just added an error when a sample has a gene panel for mutation profile and is not in the `_sequenced` case list. (#5602)

4. Currently fusion events should be added to the `_sequenced` case list. They can move to their own case list when fusions are moved the structural variant data format. Which then also should have a "#structural variant samples:" comment line.
